### PR TITLE
[AM][OPS-10831] review logging (levels, recoding of context, key logs)

### PR DESCRIPTION
**Pull request by almorcrette**

### CODE REVIEW

File: app/langswitch/LanguageSwitchController.scala [hunk 1]
```diff
@@ -30,11 +31,13 @@ class LanguageSwitchController @Inject() (
     cc:             ControllerComponents)(
     implicit
     ec: ExecutionContext)
-  extends AbstractController(cc) {
+  extends AbstractController(cc) with Logging {
 
   import requestSupport._
 
   def switchToLanguage(language: Language): Action[AnyContent] = cc.actionBuilder { implicit request =>
+    appLogger.info("Switch language")
```
Comment (AndyHWChung) [inline app/langswitch/LanguageSwitchController.scala]: having a log every time we hit a controller endpoint seems to be a bit excessive - we should get those logs automatically anyway. You can see the automatics logs if you search by session id in kibana

File: app/bars/BarsConnector.scala [hunk 2]
```diff
@@ -66,11 +66,13 @@ class BarsConnector @Inject() (
               if (barsError.code == BarsError.sortCodeOnDenyList) {
                 BarsResponseSortCodeOnDenyList(barsError)
               } else {
-                val msg = new RuntimeException(s"Unhandled error code for 400 HttpResponse: [$barsError]")
-                logger.error("couldn't validateBank", msg)
-                throw new RuntimeException(msg)
+                val ex = new RuntimeException(s"Unhandled error code for 400 HttpResponse: [$barsError]")
```
Comment (AndyHWChung) [inline app/bars/BarsConnector.scala]: if this exception were just thrown here then you'd automatically get a stack trace that you can see in kibana - we probbaly don't need additional logging here cos the exception message and the stack trace should be enough info

File: app/util/DataLogging.scala [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    def debug(message: => String, arrangement: TTPArrangement)(implicit request: RequestHeader): Unit = logMessage(message, arrangement, Debug)
+
+    def info(message: => String, arrangement: TTPArrangement)(implicit request: RequestHeader): Unit = logMessage(message, arrangement, Info)
+
+    def warn(message: => String, arrangement: TTPArrangement)(implicit request: RequestHeader): Unit = logMessage(message, arrangement, Warn)
+
+    def error(message: => String, arrangement: TTPArrangement)(implicit request: RequestHeader): Unit = logMessage(message, arrangement, Error)
+
+    private def appendedArrangement(arrangement: TTPArrangement): String = s" [arrangement: ${appendedData(Json.toJson(arrangement.obfuscate))}]"
```
Comment (AndyHWChung) [inline app/util/DataLogging.scala]: here and below, I'm not sure we really to ever log details of an arrangement - if we need the info we can look in splunk. We are obfuscating here but it feels like the kibana logs shouldn't have this level of detail in it.

File: app/util/Logger.scala [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+  def info(message: => String, ex: Throwable)(implicit request: RequestHeader): Unit = logMessage(message, ex, Info)
+
+  def warn(message: => String, ex: Throwable)(implicit request: RequestHeader): Unit = logMessage(message, ex, Warn)
+
+  def error(message: => String, ex: Throwable)(implicit request: RequestHeader): Unit = logMessage(message, ex, Error)
+
+  protected def source = s"[class: $inClass]"
+
+  protected def context(implicit rh: RequestHeader) = s"$request $sessionId $requestId $referer $deviceId"
```
Comment (AndyHWChung) [inline app/util/Logger.scala]: this seems excessive - we get this info automatically anyway as other fields in kibana logs

File: app/langswitch/LanguageSwitchController.scala (hunk 1 above)
Comment (almorcrette) [reply app/langswitch/LanguageSwitchController.scala]: Ah yes, ok. Some of the auto logging isn't currently running in production, but will be with deployment of main an `bootstrap-play` 7.15.0 and beyond where the default logging module is enabled by default.

File: app/bars/BarsConnector.scala (hunk 2 above)
Comment (almorcrette) [reply app/bars/BarsConnector.scala]: removed manual logging

File: app/util/DataLogging.scala (hunk 3 above)
Comment (almorcrette) [reply app/util/DataLogging.scala]: Ok. Currently we would log the obfuscated arrangement if the arrangement set up didn't work (i.e. the DD instruction has been successful but arrangement hasn't gone through and Debt Management would then have to manually set up the arrangement). The arrangement included the payment plan reference and the direct debit reference.

We also audit this and the audit also has the payment plan reference and direct debit reference.

This allows us to cross reference the audit event and the log.

If we remove the arrangement from the log, without including the payment plan reference and/or direct debit reference, we won't be able to cross check the logs against the audits.

Is that fine? Or shall I include the references in the log message?

File: app/util/Logger.scala (hunk 4 above)
Comment (almorcrette) [reply app/util/Logger.scala]: removed `sessionId` and `requestId` which are kibana log fields. Removed `deviceId` as excessive. Let `request` and `referer` as not in kibana log fields (that are in the request fields but useful to have them visible in the logs

File: app/langswitch/LanguageSwitchController.scala (hunk 1 above)
Comment (almorcrette) [reply app/langswitch/LanguageSwitchController.scala]: Have removed logging that only capture the controller endpoint hits (where no journey information is imparted).

Would you like me to also remove all the logging which capture the controller endpoint hits with journeyId info? These are currently only at INFO level so won't make it to production kibana anyway.

I find them fairly helpful to track the logging in the terminal when I run the app locally, but they are a bit of a burden to maintain the convention in the code.

Comment (almorcrette) [reply app/langswitch/LanguageSwitchController.scala]: I have decided to leave in logging of controller end point calls with journeyId info at INFO level as I find them useful when running the service locally. They will not appear in production as we are setting the level there to WARN.

File: app/util/DataLogging.scala (hunk 3 above)
Comment (almorcrette) [reply app/util/DataLogging.scala]: I have decided to remove the arrangement from the logging but to include the payment plan reference and direct debit reference for ease of cross referencing with splunk for the purposes to investigation if issues arise.

Comment (almorcrette) [reply app/util/DataLogging.scala]: I've now removed the payment plan reference as I realised it was the UTR, thus PII.

**Review by AndyHWChung — APPROVED** [review]

**Pull request merged**